src/server/qgsconfigcache.h

```diff
@@ -65,20 +66,46 @@ class SERVER_EXPORT QgsConfigCache : public QObject
      */
     const QgsProject *project( const QString &path, const QgsServerSettings *settings = nullptr );
 
+    /**
+     * Set the invalidation check interval
+     *
+     * If \a msec is set to a positive value then the cache will check for invalidation
+     * every \a msec milliseconds
+     *
+     * If \a msec is set to 0 (or negative value), which is the default, (or negative value) then the cache will use the QFileSystemWatcher for
+     * invalidating cache
+     *
+     * \since QGIS 3.24
+     */
+    void setCheckInterval( int msec );
```

Comment:
I might be missing the point here, but I would like to see a separate logic to control the filesystem watcher and the new timer-based watcher.
Am I correct that given this implementation the two are mutually exclusive?
I see a use case where there is a mix of filesystem-based projects and other storages, but the timer-base approach would not work with the filesystem-based projects.

Reply:
> Am I correct that given this implementation the two are mutually exclusive?

Yes

> but the timer-base approach would not work with the filesystem-based projects.

Why do you think it wont't work ? The timer approach is  based is based on the 'QgsProject::lasteModified()` call which handle file based projects as well as other storage.

Since the timer strategy handle files the two methods will probably walk on each other steps.

Reply:
> > Am I correct that given this implementation the two are mutually exclusive?
> 
> Yes
> 
> > but the timer-base approach would not work with the filesystem-based projects.
> 
> Why do you think it wont't work ? The timer approach is based is based on the 'QgsProject::lasteModified()` call which handle file based projects as well as other storage.

The lastModified call calls readProjectStorageMetadata which is not implemented for filesystem-based projects and even if it was implemented it would need to reload and parse the project's XML which is one of the slow steps here: we definitely want to avoid that unless it's implemented in a separate thread.

Reply:
On which level should this be configurable? Per project? Path prefix?

I think the current approach suits many cases (actually I am not aware of an installation where we would mix different storage types with one server instance) and it's quite simple, which speaks for it in terms of expected problems and maintenance overhead.

My approach would be to offer some simple strategies (as done here) plus an option to disable this completely and implement dedicated strategies via custom logic (plugin).

Reply:
> The lastModified call calls readProjectStorageMetadata which is not implemented for filesystem-based projects

AFAIK, the lastModified() handle files by calling directly `QFileInfo::lastModified`: https://qgis.org/api/qgsproject_8cpp_source.html#l00686

> It would need to reload and parse the project's XML which is one of the slow step

Why it would need to do that ? this is a simple modified time check on a file.

Reply:
> Why it would need to do that ? this is a simple modified time check on a file.

You are right: I thought there was a filesystem storage implementation which is not the case.

Reply:
Note that the tests ares based on files :-)